harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
+
+    def _LRScheduler(
+        self,
+        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: pytorch.LRScheduler.StepMode,
+    ) -> pytorch.LRScheduler:
+        # Initialize LR schedulers. This block must happen after wrapping the optimizers using
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
+        self.lr_schedulers.append(wrapped)
+        return wrapped
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if self.hvd_config.use:
+            check.gt(self.n_gpus, 0)
+            # We launch a horovod process per GPU. Each process
+            # needs to bind to a unique GPU.
+            self.device = torch.device(hvd.local_rank())
+            torch.cuda.set_device(self.device)
+        elif self.n_gpus > 0:
+            self.device = torch.device("cuda", 0)
+        else:
+            self.device = torch.device("cpu")
+        check.is_not_none(self.device)
+
+    def _to_device(self, data: pytorch._Data) -> pytorch.TorchData:
+        return pytorch.to_device(
+            data, self.device, self.warning_logged[_WarningLogs.FAILED_MOVING_TO_DEVICE]
+        )
+
+    def _use_amp(self) -> bool:
+        return self.mixed_precision != "O0"
+
+    def _configure_amp(
+        self,
+        model: torch.nn.Module,
+        optimizers: Union[torch.optim.Optimizer, List[torch.optim.Optimizer]],  # type: ignore
+        mixed_precision: str = "O0",
+    ) -> Tuple[torch.nn.Module, List[torch.optim.Optimizer]]:  # type: ignore
+        self.mixed_precision = mixed_precision
```

Comment:
blocking: checking mixed precision setting should all be done in `_use_amp()` otherwise you get ordering requirements between the functions.

Reply:
On second thought, since users will be calling `_configure_amp()` themselves, we should get rid of `_use_amp()` function, and instead have a private bool `self._using_amp` that gets set to `True` if `_configure_amp()` is called. We also should get rid of the  `if mixed_precision != "O0":` check below, if a user wants to configure amp with "O0" we should let them. However, until we switch over to the new API, we should call `_configure_amp()` ourselves only when `ExperimentConfig.mixed_precision_enabled()` is `True`.

Reply:
So if we allow setting `O0` in mixed precision not take it as mixed precision unset. We need to change the master's logic of setting mixed_precision as default. This involves a large change in the whole cluster. I'd leave to #ml-ag to figure this out.

Reply:
Done. But I wouldn't use `ExperimentConfig.mixed_precision_enabled()` since it will be removed along with supporting AMP in the constructor.